openedx/xblock-lti-consumer
# Fix PII Sharing Behavior and Enable PII Sharing in LTI 1.3 Launches and Fix LTI 1.3 Modal Launches

### Description

This pull request fixes a few issues and bugs with sharing personally identifiable information (PII) in LTI launches as well as one bug with the `modal` LTI launch in LTI 1.3.

The follow issues and bugs were resolved in this pull request:
* `username` and `email` are not sent in LTI 1.3 launches
* the PII sharing LTI XBlock fields always appear, regardless of the existence and value of a `CourseAllowPIISharingInLTIFlag`
* `inline` LTI launches did not trigger a PII consent modal
* `modal` LTI launches did not trigger a PII consent modal
* the `modal` LTI launch for LTI 1.3 failed due to differences in the form of the launch URL provided to the Javascript via the XBlock code
* the values returned by `lti_access_to_learners_editable` were not correct, and the values were not being recomputed on each request

Please see the `CHANGELOG` for a more detailed description of the bugs and their fixes.

### UI/UX Changes

The PII consent modal was replaced with an inline consent dialog for all LTI launch types and for both LTI 1.1 and LTI 1.3.

**LTI 1.1**

![MST-1736-Demo-LTI-PII-Consent-Dialog-LTI-1 1](https://user-images.githubusercontent.com/11871801/207914499-cb5310b7-f440-4f79-9f85-b3d8fcf5fc3b.gif)

**LTI 1.3**

![MST-1736-Demo-LTI-PII-Consent-Dialog-LTI-1 3-Inline](https://user-images.githubusercontent.com/11871801/207915820-ad03838e-e111-4ed4-af55-eec308e2ead8.gif)
![MST-1736-Demo-LTI-PII-Consent-Dialog-LTI-1 3-Modal](https://user-images.githubusercontent.com/11871801/207915836-324f56a6-d312-4720-9469-1b8a1c682fbf.gif)
![MST-1736-Demo-LTI-PII-Consent-Dialog-LTI-1 3-New-Window](https://user-images.githubusercontent.com/11871801/207915850-1a0b4b8e-c559-40ed-afd5-8f46a16a501c.gif)

### Manual Testing

These changes were tested on all three LTI launch types (i.e. `inline`, `modal` and `new_window`) across both LTI 1.1 and LTI 1.3.

* I tested that the `CourseAllowPIISharingInLTIFlag` controls the appearance of the PII sharing LTI XBlock fields.
* I tested that the consent dialog appears for all three LTI launch types and works as expected.
* I tested that the consent dialog appears only when any or all of the PII sharing LTI XBlock fields is set to True.
* I tested that the backward compatibility code in `lti_access_to_learners_editable` works correctly. Instances of `CourseAllowPIISharingInLTIFlag` are created successfully when appropriate.
* I tested that `lti_access_to_learners_editable` returns the correct value.
* I tested that `username` and `email` are transmitted in LTI 1.3 launches when appropraite.
